### PR TITLE
Salt Cloud fixes

### LINKED ISSUE #34033: salt-cloud throws an AttributeError when trying to delete nodes via a map file with a non-existant profile
### Description of Issue/Question

`salt-cloud` is throwing an AttributeError when trying to remove nodes defined in a map file when a profile does not exist. 
### Steps to Reproduce Issue
1. Create a map file
2. Create nodes using `salt-cloud -m <your_map>`
3. Remove all nodes using `salt-cloud -m <your_map> -d`

**Important:** using vultr as a provider but I don't think it really matters.

**Trace:**

```
> /usr/lib/python2.7/site-packages/salt/cloud/__init__.py(1713)delete_map()
   1712         for alias, drivers in six.iteritems(query_map.copy()):
-> 1713             for driver, vms in six.iteritems(drivers.copy()):
   1714                 for vm_name, vm_details in six.iteritems(vms.copy()):

ipdb> alias
ipdb> drivers
{'x': "No provider for the mapped 'x' profile was found. Skipped VMS: x"}
```

While iterating over the next line:

```
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
AttributeError: 'str' object has no attribute 'copy'
Traceback (most recent call last):
  File "/usr/bin/salt-cloud", line 10, in <module>
    salt_cloud()
  File "/usr/lib/python2.7/site-packages/salt/scripts.py", line 435, in salt_cloud
    client.run()
  File "/usr/lib/python2.7/site-packages/salt/cloud/cli.py", line 191, in run
    matching = mapper.delete_map(query='list_nodes')
  File "/usr/lib/python2.7/site-packages/salt/cloud/__init__.py", line 1713, in delete_map
    for vm_name, vm_details in six.iteritems(vms.copy()):
AttributeError: 'str' object has no attribute 'copy'
Traceback (most recent call last):
  File "/usr/bin/salt-cloud", line 10, in <module>
    salt_cloud()
  File "/usr/lib/python2.7/site-packages/salt/scripts.py", line 435, in salt_cloud
    client.run()
  File "/usr/lib/python2.7/site-packages/salt/cloud/cli.py", line 191, in run
    matching = mapper.delete_map(query='list_nodes')
  File "/usr/lib/python2.7/site-packages/salt/cloud/__init__.py", line 1713, in delete_map
    for vm_name, vm_details in six.iteritems(vms.copy()):
AttributeError: 'str' object has no attribute 'copy'
```
### Versions Report

```
Salt Version:
           Salt: 2016.3.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.5 (default, Nov 20 2015, 02:00:19)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.7.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: centos 7.2.1511 Core
        machine: x86_64
        release: 3.10.0-327.18.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.2.1511 Core
```

### LINKED ISSUE #56589: `salt-cloud` failing to list nodes for recent versions of `apache-libcloud`
### Description of Issue

The `node_state` function in `libcloudfuncs.py` is throwing errors related to the `id_` input type:

```
[DEBUG   ] Failed to execute 'gce.list_nodes()' while querying for running nodes: running
Traceback (most recent call last):
  File "/home/br/Dev/f0cal/2020-04/_venv/lib/python3.6/site-packages/salt/cloud/__init__.py", line 2366, in run_parallel_map_providers_query
    cloud.clouds[data['fun']]()
  File "/home/br/Dev/f0cal/2020-04/_venv/lib/python3.6/site-packages/salt/cloud/libcloudfuncs.py", line 467, in list_nodes
    'state': node_state(node.state)
  File "/home/br/Dev/f0cal/2020-04/_venv/lib/python3.6/site-packages/salt/cloud/libcloudfuncs.py", line 86, in node_state
    return states_str[id_] if isinstance(id_, six.string_types) else states_int[id_]
KeyError: running
```

A quick look at [closed issues](https://github.com/saltstack/salt/issues?q=is%3Aissue+node_state+sort%3Aupdated-desc+) reveals that we've seen this before.

First, `libcloud` was passing a `node.state` as an `int` and then a `str` after v1.0. Now (v2.8.1) it appears to be passing a `NodeState` object that needs to be [coerced](https://github.com/apache/libcloud/blob/3086851e1c7f5838cd9c0fb42a31bdf08b6f9706/libcloud/compute/types.py#L44).

### Setup
N/A

### Steps to Reproduce Issue
Run `salt-cloud` with recent version of `apache-libcloud` and debug on.

### Versions Report

```
Salt Version:
           Salt: 3000.1
 
Dependency Versions:
           cffi: 1.14.0
       cherrypy: Not Installed
       dateutil: 2.8.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.11.1
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: 2.20
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.9 (default, Nov  7 2019, 10:44:02)
   python-gnupg: Not Installed
         PyYAML: 5.3.1
          PyZMQ: 19.0.0
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.2
 
System Versions:
           dist: Ubuntu 18.04 bionic
         locale: UTF-8
        machine: x86_64
        release: 4.15.0-91-generic
         system: Linux
        version: Ubuntu 18.04 bionic
```

### LINKED ISSUE #56862: [BUG] `salt-cloud` Failed to get the output of 'gce.avail_images()'
**Description**
``` salt-cloud -l debug --list-images my-google``` returns:

```
[ERROR   ] Failed to get the output of 'gce.avail_images()': {u'reason': u'notFound', u'message': u"The resource 'projects/c' was not found. A missing project error may be an authentication issue. Please  ensure your auth credentials match your project. ", u'domain': u'global'}
```

**Setup**
```
python -c "import libcloud ; print libcloud.__version__"
2.2.1
```
`/etc/salt/cloud.providers.d/google.conf`
```
my-google:
  project: 'my-project'
  service_account_email_address: ''
  service_account_private_key: ''
  minion:
    master: tmp-salt-master-03.example.com
  driver: gce
```

**Steps to Reproduce the behavior**

On a Ubuntu 18.04 minimal with full API access on us-east4-a

```
sudo apt-get update
sudo apt-get install tmux git emacs-nox vim -y

curl -o bootstrap-salt.sh -L https://bootstrap.saltstack.com
tmux
sudo sh bootstrap-salt.sh git develop

sudo su

mkdir -p /etc/salt/cloud.providers.d

cat <<EOT >>  /etc/salt/cloud.providers.d/google.conf
my-google:
  project: 'my-project'
  service_account_email_address: ''
  service_account_private_key: ''
  minion:
    master: tmp-salt-master-03.example.com
  driver: gce
EOT

sudo apt-get install python-libcloud -y
```

`salt-cloud -l debug --list-images my-google`
```
WARNING: CONFIG '/etc/salt/cloud' file does not exist. Falling back to default '/etc/salt/cloud'.
[DEBUG   ] Missing configuration file: /etc/salt/cloud
[DEBUG   ] Missing configuration file: /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: tmp-salt-master-02.c.covidpr-271900.internal
[DEBUG   ] Missing configuration file: /etc/salt/cloud.providers
[DEBUG   ] Including configuration from '/etc/salt/cloud.providers.d/google.conf'
[DEBUG   ] Reading configuration from /etc/salt/cloud.providers.d/google.conf
[DEBUG   ] Missing configuration file: /etc/salt/cloud.profiles
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[INFO    ] salt-cloud starting
[DEBUG   ] Could not LazyLoad parallels.avail_sizes: 'parallels' __virtual__ returned False
[DEBUG   ] LazyLoaded parallels.avail_locations
[DEBUG   ] LazyLoaded proxmox.avail_sizes
[DEBUG   ] Starting new HTTP connection (1): metadata
[DEBUG   ] http://metadata:80 "GET /computeMetadata/v1/ HTTP/1.1" 200 28
[DEBUG   ] Starting new HTTPS connection (1): www.googleapis.com
[DEBUG   ] https://www.googleapis.com:443 "GET /compute/v1/projects/covidpr-271900/zones HTTP/1.1" 200 None
[DEBUG   ] https://www.googleapis.com:443 "GET /compute/v1/projects/covidpr-271900/regions HTTP/1.1" 200 None
[DEBUG   ] https://www.googleapis.com:443 "GET /compute/v1/projects/c/global/images HTTP/1.1" 404 None
[ERROR   ] Failed to get the output of 'gce.avail_images()': {u'reason': u'notFound', u'message': u"The resource
 'projects/c' was not found. A missing project error may be an authentication issue. Please  ensure your auth cr
edentials match your project. ", u'domain': u'global'}
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 866, in image_list
    data[alias][driver] = self.clouds[fun]()
  File "/usr/lib/python2.7/dist-packages/salt/cloud/clouds/gce.py", line 363, in avail_images
    all_images.extend(conn.list_images(project))
  File "/usr/lib/python2.7/dist-packages/libcloud/compute/drivers/gce.py", line 2217, in list_images
    ex_include_deprecated=dep)
  File "/usr/lib/python2.7/dist-packages/libcloud/compute/drivers/gce.py", line 2269, in ex_list_project_images
    method='GET').object
  File "/usr/lib/python2.7/dist-packages/libcloud/compute/drivers/gce.py", line 121, in request
    response = super(GCEConnection, self).request(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/libcloud/common/google.py", line 808, in request
    *args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/libcloud/common/base.py", line 637, in request
    response = responseCls(**kwargs)
  File "/usr/lib/python2.7/dist-packages/libcloud/common/base.py", line 159, in __init__
    self.object = self.parse_body()
  File "/usr/lib/python2.7/dist-packages/libcloud/common/google.py", line 280, in parse_body
    raise ResourceNotFoundError(message, self.status, code)
ResourceNotFoundError: {u'reason': u'notFound', u'message': u"The resource 'projects/c' was not found. A missing
 project error may be an authentication issue. Please  ensure your auth credentials match your project. ", u'dom
ain': u'global'}
[DEBUG   ] LazyLoaded nested.output
```

**Expected behavior**
A List of available images instead of an error

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
alt Version:
           Salt: 2019.8.0-69-g637fe0b
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: 0.27.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.17 (default, Apr 15 2020, 17:20:14)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5
 
System Versions:
           dist: Ubuntu 18.04 bionic
         locale: UTF-8
        machine: x86_64
        release: 5.0.0-1034-gcp
         system: Linux
        version: Ubuntu 18.04 bionic
 
```
</details>

### PR DESCRIPTION
**Pull request by s0undt3ch**
### What does this PR do?
See title

---
- Fixes #34033
- Fixes #56589
- Fixes #56862

### CODE REVIEW

**Review by krionbsd — DISMISSED** [review]

**Comment by bryceml** [comment]
@s0undt3ch do we need to update libcloud then?  That's going to be a bit of work, most oses provide 2.2 I think.

**Comment by s0undt3ch** [comment]
The libcloud requirent stays at 1.5.0, however, to use GCE, 2.5.0 is required.
Authenticating against GCE simplify fails with older versions.
Seems related to the fact that we no longer use Pycrypto and libcloud stops using Pycrypto on 2.5.0

**Review by krionbsd — DISMISSED** [review]

File: salt/loader.py [hunk 1]
```diff
@@ -2197,14 +2197,14 @@ def _process_virtual(self, mod, module_name, virtual_func="__virtual__"):
 
         return (True, module_name, None, virtual_aliases)
 
-    def run(self, method, *args, **kwargs):
+    def run(self, _func_or_method, *args, **kwargs):
```
Comment (Ch3LL) [inline salt/loader.py:2200]: Why is this change necessary? I'm concerned we are changing the arg names on a publicly accessible method.

File: requirements/static/ci/common.in [hunk 2]
```diff
@@ -2,7 +2,7 @@
 # We can also exclude platforms from the requirements using markers, but if a requirement only applies
 # to a particular platform, please add it to the corresponding `<platform>.in` file in this directory.
 azure==4.0.0; sys_platform != 'win32'
-apache-libcloud>=2.4.0; sys_platform != 'win32'
+apache-libcloud>=1.5.0; sys_platform != 'win32'
```
Comment (Ch3LL) [inline requirements/static/ci/common.in:5]: Why are you downgrading the version here?

**Review by Ch3LL — CHANGES_REQUESTED** [review]
I do not think we are running joyent integration tests anymore. If so can we get some test coverage for that cloud provider since there was multiple changes.

File: salt/loader.py (hunk 1 above)
Comment (s0undt3ch) [reply salt/loader.py:2200]: These functions are the loader changes from @dwoz, unreleased.
They needed to be named more uniquely because if the target function also had a `method` argument, python would blow up.

File: requirements/static/ci/common.in (hunk 2 above)
Comment (s0undt3ch) [reply requirements/static/ci/common.in:5]: We're only downgrading the **minimum** requirement, the actual requirement versions are the same.
@bryceml mentioned that for some distros, we still package 1.5.0, hence this change.

**Comment by s0undt3ch** [comment]
> I do not think we are running joyent integration tests anymore. If so can we get some test coverage for that cloud provider since there was multiple changes.

What exactly do you mean? Write unit tests for joyent?

Also, the PR started as just getting cloud tests running again. Additional test coverage should be done on a separate PR, agree?

**Comment by Ch3LL** [comment]
Yes, unit tests. There is joyent unit tests currently but would be good to add one for the new change with node state (for example this method: joyent_node_state). I'm fine adding the support in a separate PR.

**Review by Ch3LL — DISMISSED** [review]

**Comment by s0undt3ch** [comment]
FYI, no new functionality was added to joyent. It was relying on a function from libcloud which no longer makes sense, but since we don't run integration tests against joyent, I just copied that function to joyent in order not to break it, that was it.

**Review by garethgreenaway — DISMISSED** [review]
